### PR TITLE
Integrate market creator fees into orderbook

### PR DESCRIPTION
**Pull request by Chralt98**
### What does it do?

Fixes https://github.com/zeitgeistpm/zeitgeist/issues/1113

Implement market creator fees for orderbook. The fees are always taken in the base asset of the market.

### CODE REVIEW

**Review by maltekliemann — CHANGES_REQUESTED** [review]
I think we should take fees only when orders are filled. Otherwise, adjusting prediction will quickly becomes too expensive.

File: primitives/src/traits/distribute_fees.rs [hunk 1]
```diff
@@ -22,6 +22,8 @@ pub trait DistributeFees {
     type Balance;
     type MarketId;
 
+    fn get_fee(market_id: Self::MarketId, amount: Self::Balance) -> Self::Balance;
```
[the remark this answers is not in the document]
Comment (Chralt98) [reply primitives/src/traits/distribute_fees.rs:25]: I can only see `market_id` and `amount`, but not `asset` and `account`.

[the remark this answers is not in the document]
Comment (Chralt98) [reply primitives/src/traits/distribute_fees.rs:25]: Ok, saw your comment below...

[the remark this answers is not in the document]
Comment (Chralt98) [reply primitives/src/traits/distribute_fees.rs:34]: I can do this, but why should we currently change it, if it is not required? I mean the API can adapt at any time if those parameters are needed.

[the remark this answers is not in the document]
Comment (maltekliemann) [reply primitives/src/traits/distribute_fees.rs:32]: `distribute` allows the amount of distribution to depend on `asset` and `account`. The API should reflect that. Since you're introducing `get_fee` here, this is how I'd do it.

File: zrml/orderbook/src/lib.rs [hunk 2]
```diff
@@ -386,34 +411,52 @@ mod pallet {
...
+                        outcome_asset,
+                        base_asset,
+                        // only outcome_asset_amount_minus_fees left to repatriate
+                        outcome_asset_amount: outcome_asset_amount_minus_fees,
+                        // maker requests full base asset amount (no fees charged yet) from taker(s)
+                        base_asset_amount,
+                    }
                 }
-            }
+            };
```
Comment (maltekliemann) [inline zrml/orderbook/src/lib.rs:459]: Seems to me like this isn't what we want.

1. Alice wants to sell 100 units of A. But for some reason, she's only selling 99. The last unit just sits there on here account. Why?
2. If we calculate the price as `base_asset_amount / outcome_asset_amount`, then the price will be wrong (too high).

I suggest to revert these changes.

**Review by maltekliemann — CHANGES_REQUESTED** [review]

File: zrml/orderbook/src/lib.rs (hunk 2 above)
Comment (Chralt98) [reply zrml/orderbook/src/lib.rs:459]: I've been passionately waiting for this discussion :-) thanks for bringing this to our focus now!

The reasoning behind this was that both parties, taker and maker should pay their fair share of the trade.

For this to be equal, we charge the fee from the `base_asset_amount` in the `fill_order` function, resulting in fees for the one who gets the base asset. However, the one who gets the outcome asset should also pay their fair share (creator fees), resulting in the same percentage share as charged for the base asset, but now for the outcome asset. 

Do I missing something or is my calculation right? But I totally agree, it's confusing, because the maker did specify an exact `outcome_asset_amount` and actually is charged with less. So I would love to avoid this confusing point. So, I am open for suggestions here!

Comment (Chralt98) [reply zrml/orderbook/src/lib.rs:459]: > If we calculate the price as base_asset_amount / outcome_asset_amount, then the price will be wrong (too high).

The fees are always taken upfront, so the price shouldn't be changed, because we basically have:

(base_asset_amount - fees) / (outcome_asset_amount - fees)

Comment (Chralt98) [reply zrml/orderbook/src/lib.rs]: https://github.com/zeitgeistpm/zeitgeist/pull/1167/commits/56679b768b5a99727bf9d139ca74de500a8a360d

**Comment by Chralt98** [comment]
Abandoned in favor of https://github.com/zeitgeistpm/zeitgeist/pull/1183
